### DISCUSSION TITLE
Dealing with properties where null is valid

### DISCUSSION THREAD

**Discussion by speterson-zoll** [louthy/language-ext#1120] [Q&A] [answered]
I'm still learning the library, but I get stuck on things.  In my case I have a DTO class with this property:

`public string? EventDateTime { get; set; }`

For this property, a null is perfectly valid; the date is not required.  However, I still need to parse/validate the string to be sure it's a proper `DateTime`.  I wrote this parser:

```
public static Validation<Error, DateTime> ParseNullableDateTime(string? dateTime) =>
        new(notnull(dateTime)
            ? parseDateTime(dateTime).ToValidation(Error.New($"{dateTime} is not a valid Date"))
            : null);
```

The Parser works just fine for a valid date; it returns a properly parsed `DateTime`.  It also works fine for an invalid date string; it returns a `Fail` with the error message shown above.  However, the `null` business is an issue.  I get a `Fail` with an empty list, but the return value also shows a `Success` with the default `DateTime`:

```
[0] = {ValidationData<Error, DateTime>} ValidationData(Fail, 1/1/0001 12:00:00 AM, [])
 Fail = {Lst<Error>} []
 State = {Validation.StateType} Fail
 Success = {DateTime} 1/1/0001 12:00:00 AM 
```

I was hoping that a `null` would count as a `Success`, but I see now that that can't happen.

I also tried changing the `string?` type to `Option<string>` and parsing that, but I get a `Value is null` exception:

```
public static Validation<Error, DateTime?> ParseNullableDateTime(Option<string> value) =>
        value.Match(
            Some: d => DateTime.TryParse(d, out var date)
                ? Success<Error, DateTime?>(date)
                : Fail<Error, DateTime?>(Error.New($"{d} is not a valid Date")),
            None: () => null);
```

Any advice on how to validate a `string?` property to a `DateTime` is welcome.

**Comment by YuriiNskyi** [comment]
I'm not sure why do you still need to return nullable `DateTime`, please correct me, if I'm mistaken and you really need to do this.

Anyway, IMHO the clearest solution is something like this:

```c#
public static Validation<Error, DateTime> ParseNullableDateTime(string? dateTime)
    => Prelude.Optional(dateTime)
        .Bind(Prelude.parseDateTime)
        .ToValidation(Error.New($"{dateTime} is not a valid Date"))
```

This signature is kind of strange:
```c#
Validation<Error, DateTime?>
```

Forget about null when you have already valid `DateTime`, you don't need to bother with nulls anymore.

**Comment by sparerd** [comment, marked as answer]
I think you have it *almost* right with your last example. The `Value is null` exception you are getting is because LanguageExt forbids nulls unless explicitly allowed (using the `____Unsafe` functions). In this case, the `Validation` type does not have any overloads for accepting nulls.

Rather than propagating the nullable type, consider going all-in on using the alternate value monads. Changing the return type from `Validation<Error, DateTime?>` to `Validation<Error, Option<DateTime>>` provides the same information but with better safety. Now instead of a `Success` case with a null value, we give it `None`.

```csharp
public static Validation<Error, Option<DateTime>> ParseNullableDateTime(Option<string> value) =>
	value.Match(
		Some: d => DateTime.TryParse(d, out var date)
			? Success<Error, Option<DateTime>>(date)
			: Fail<Error, Option<DateTime>>(Error.New($"{d} is not a valid Date")),
		None: () => Success<Error, Option<DateTime>>(None));
```

And if you do ever *really* need a nullable DateTime, you can still `MatchUnsafe()` on your `Option<DateTime>`  and map the `None` to `null`.
